maskrcnn_benchmark/data/datasets/voc.py

```diff
@@ -0,0 +1,117 @@
+import os
+import time
+
+import torch
+import torch.utils.data as data
+from PIL import Image
+from lxml import etree
+
+from maskrcnn_benchmark.structures.bounding_box import BoxList
+
+VOC_BBOX_LABEL_NAMES = ('__background__ ',
+                        'aeroplane',
+                        'bicycle',
+                        'bird',
+                        'boat',
+                        'bottle',
+                        'bus',
+                        'car',
+                        'cat',
+                        'chair',
+                        'cow',
+                        'diningtable',
+                        'dog',
+                        'horse',
+                        'motorbike',
+                        'person',
+                        'pottedplant',
+                        'sheep',
+                        'sofa',
+                        'train',
+                        'tvmonitor')
+
+
+class PascalVOC(data.Dataset):
+    def __init__(self, data_dir, split, name, transforms=None):
+        self.data_dir = data_dir
+        self.split = split
+        self.name = name
+
+        self.anns_dir = os.path.join(data_dir, 'Annotations')
+        self.imgs_dir = os.path.join(data_dir, 'JPEGImages')
+        split_file = os.path.join(data_dir, 'ImageSets', 'Main', '%s.txt' % split)
+
+        print('loading %s annotations into memory...' % name)
+        tic = time.time()
+        with open(split_file) as fid:
+            self.ids = sorted([l.strip() for l in fid.readlines()])
+        self.anns = {}
+        for img_id in self.ids:
+            with open(os.path.join(self.anns_dir, '%s.xml' % img_id)) as fid:
+                xml_str = fid.read()
+            xml = etree.fromstring(xml_str)
+            data = self._recursive_parse_xml_to_dict(xml)['annotation']
+            self.anns[img_id] = {
+                'boxes': [[int(obj['bndbox']['xmin']),
+                           int(obj['bndbox']['ymin']),
+                           int(obj['bndbox']['xmax']),
+                           int(obj['bndbox']['ymax'])] for obj in data['object'] if int(obj['difficult']) == 0],
```

Comment:
I think this is not the best thing to do unconditionally.

Indeed, we might want to remove the difficult instances during training, but during testing there should be a way of knowing which one of the instances are difficult, so that they are not taken into account during evaluation.

If I understand it correctly, the way it is currently done will be penalizing the difficult instances, as if we detect a difficult instance with our model, it will be counted as a false match.

I think we should have a flag that lets us decide if we want to return difficult instances or not.

Reply:
I got it.